Self-Consistent-Field/vcd.py

```diff
@@ -0,0 +1,785 @@
+
+for key in Hes:
+    Mat = psi4.core.Matrix.from_array(Hes[key])
+    if key in oei_dict:
+        Mat.name = oei_dict[key] + " HESSIAN"
+        Mat.print_out()
+        psi4.core.print_out("\n")
+
+
+for atom1 in range(natoms):
+    for atom2 in range(natoms):
+        string = "TEI" + str(atom1) + str(atom2)
+        deriv2_mat[string] = mints.mo_tei_deriv2(atom1, atom2, C, C, C, C)
+        pq = 0
+        for p in range(3):
+            for q in range(3):
+                map_key = string + cart[p] + cart[q]
+                deriv2_np[map_key] = np.asarray(deriv2_mat[string][pq])
+                pq = pq + 1
+                row = 3 * atom1 + p
+                col = 3 * atom2 + q
+
+                Hes["TEI"][row][col] +=  2.0 * np.einsum("pq,pqmm->", ref_opdm, deriv2_np[map_key][:, :, :nocc, :nocc], optimize=True)
+                Hes["TEI"][row][col] += -1.0 * np.einsum("pq,pmmq->", ref_opdm, deriv2_np[map_key][:, :nocc, :nocc, :], optimize=True)
+
+                Hes["TEI"][row][col] +=  2.0 * np.einsum("pq,pqmm->", Ppq-ref_opdm, deriv2_np[map_key][:, :, :nocc, :nocc], optimize=True)
+                Hes["TEI"][row][col] += -1.0 * np.einsum("pq,pmmq->", Ppq-ref_opdm, deriv2_np[map_key][:, :nocc, :nocc, :], optimize=True)
+
+                Hes["TEI"][row][col] += np.einsum("pqrs,prqs->", Ppqrs, deriv2_np[map_key], optimize=True)
+
+TEIMat = psi4.core.Matrix.from_array(Hes["TEI"])
+TEIMat.name = " TEI HESSIAN"
+TEIMat.print_out()
+
+
+# Solve the first-order CPHF equations here,  G_aibj Ubj^x = Bai^x (Einstein summation),
+# where G is the electronic hessian,
+# G_aibj = delta_ij * delta_ab * epsilon_ij * epsilon_ab + 4 <ij|ab> - <ij|ba> - <ia|jb>,
+# where epsilon_ij = epsilon_i - epsilon_j, (epsilon -> orbital energies),
+# x refers to the perturbation, Ubj^x are the corresponsing CPHF coefficients
+# and Bai^x = Sai^x * epsilon_ii - Fai^x + Smn^x  * (2<am|in> - <am|ni>),
+# where, S^x =  del(S)/del(x), F^x =  del(F)/del(x).
+
+B = {}
+U = {}
+# Build Bai^x
+for atom in range(natoms):
+    for p in range(3):
+        key = str(atom) + cart[p]
+        B[key] =  np.einsum("ai,ii->ai", deriv1_np["S" + key][nocc:, :nocc], F[:nocc, :nocc], optimize=True)
+        B[key] -= F_grad[key][nocc:, :nocc]
+        B[key] +=  2.0 * np.einsum("amin,mn->ai", npERI[nocc:, :nocc, :nocc, :nocc], deriv1_np["S" + key][:nocc, :nocc], optimize=True)
+        B[key] += -1.0 * np.einsum("amni,mn->ai", npERI[nocc:, :nocc, :nocc, :nocc], deriv1_np["S" + key][:nocc, :nocc], optimize=True)
+
+        # Compute U^x, where
+        # U_ij^x = - 1/2 S_ij^a
+        # U_ai^x = G^(-1)_aibj * B_bj^x
+        # U_ia^x = - (U_ai^x + S_ai^x)
+        # U_ab^x = - 1/2 S_ab^a
+        U[key] = np.zeros((nmo, nmo))
+        U[key][:nocc, :nocc] = - 0.5 * deriv1_np["S" + key][:nocc, :nocc]
+        U[key][nocc:, :nocc] = np.einsum("iajb,bj->ai", Ginv, B[key], optimize=True)
```

Comment:
Would it be possible to use [`np.linalg.solve`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html#numpy.linalg.solve) here instead of explicitly inverting `G`? You'd avoid storing both `G` and its inverse and save some memory.